File: golang/internal/arbitrage/calculator.go
package arbitrage

type ProfitBreakdown struct {
    GrossProfit   float64
    NetProfit     float64
    Fees          float64
    SlippageCost  float64
}

type Calculator struct {
    FeeRate  float64
    Slippage float64
}

func NewCalculator() *Calculator {
    return &Calculator{FeeRate: 0.001, Slippage: 0.0005}
}

func (c *Calculator) Evaluate(buyPrice, sellPrice, amount float64) ProfitBreakdown {
    gross := (sellPrice - buyPrice) * amount
    notional := (sellPrice + buyPrice) * amount
    fees := notional * c.FeeRate
    slippage := notional * c.Slippage
    net := gross - fees - slippage
    return ProfitBreakdown{
        GrossProfit:  round(gross, 4),
        NetProfit:    round(net, 4),
        Fees:         round(fees, 4),
        SlippageCost: round(slippage, 4),
    }
}

File: src/arbitrage/calculator.py
"""Profit estimation utilities for potential arbitrage trades."""
from __future__ import annotations

from dataclasses import dataclass

from ..utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class ProfitBreakdown:
    gross_profit: float
    net_profit: float
    fees: float
    slippage_cost: float


class OpportunityCalculator:
    """Perform simple profit calculations for arbitrage opportunities."""

    def __init__(self, fee_rate: float = 0.001, slippage: float = 0.0005) -> None:
        self.fee_rate = fee_rate
        self.slippage = slippage

    def evaluate(self, buy_price: float, sell_price: float, amount: float) -> ProfitBreakdown:
        logger.debug(
            "Evaluating opportunity buy_price=%s sell_price=%s amount=%s",
            buy_price,
            sell_price,
            amount,
        )
        gross_profit = (sell_price - buy_price) * amount
        notional = (buy_price + sell_price) * amount
        fees = notional * self.fee_rate
        slippage_cost = notional * self.slippage
        net_profit = gross_profit - fees - slippage_cost
        return ProfitBreakdown(
            gross_profit=round(gross_profit, 4),
            net_profit=round(net_profit, 4),
            fees=round(fees, 4),
            slippage_cost=round(slippage_cost, 4),
        )
